brainiak/utils/utils.py

```diff
@@ -773,25 +774,58 @@ def p_from_null(X, two_sided=False):
         above the null) or two-sided (testing for both significantly positive
         and significantly negative values)
 
+    max_null_input
+    min_null_input
+        ndarray with num_perm (see isfc.py) entries.
+        by default these arrays are derived from
+        the X input array, which can be very large
+        and takes up huge memory space.
+
     Returns
     -------
     p : ndarray the same shape as X, without the last dimension
         p values for each true X value under the null distribution
     """
-    leading_dims = tuple([int(d) for d in np.arange(X.ndim - 1)])
-
-    # Compute maximum/minimum in each null dataset
-    max_null = np.max(X[..., 1:], axis=leading_dims)
-    min_null = np.min(X[..., 1:], axis=leading_dims)
-
-    # Compute where the true values fall on the null distribution
-    max_null_ecdf = ecdf(max_null)
-    if two_sided:
-        min_null_ecdf = ecdf(min_null)
-        p = 2 * np.minimum(1 - max_null_ecdf(X[..., 0]),
-                           min_null_ecdf(X[..., 0]))
-        p = np.minimum(p, 1)
+    if (min_null_input is None) or (max_null_input is None):
+        leading_dims = tuple([int(d) for d in np.arange(X.ndim - 1)])
+
+        # Compute maximum/minimum in each null dataset
+        max_null = np.max(X[..., 1:], axis=leading_dims)
+        min_null = np.min(X[..., 1:], axis=leading_dims)
+
+        # Compute where the true values fall on the null distribution
+        max_null_ecdf = ecdf(max_null)
+        if two_sided:
+            min_null_ecdf = ecdf(min_null)
+            p = 2 * np.minimum(1 - max_null_ecdf(X[..., 0]),
+                               min_null_ecdf(X[..., 0]))
+            p = np.minimum(p, 1)
+        else:
+            p = 1 - max_null_ecdf(X[..., 0])
+
     else:
-        p = 1 - max_null_ecdf(X[..., 0])
+        # maximum & minimum in each null dataset should be provided as input
+        max_null = max_null_input
+        min_null = min_null_input
+
+        max_null_ecdf = ecdf(max_null)
+        if two_sided:
+            min_null_ecdf = ecdf(min_null)
+            p = 2 * np.minimum(1 - max_null_ecdf(X),
+                               min_null_ecdf(X))
+            p = np.minimum(p, 1)
+        else:
+            p = 1 - max_null_ecdf(X)
 
     return p
+
+
+def gen_null_array(n_subj, n_perm, collapse_subj, float_type):
```

Comment:
This is pretty specific to the eventseg code, so I would move this to eventseg.py. It also needs a (brief) docstring, and maybe a better name to clarify that the arrays are empty rather than null samples (e.g. initialize_null_arrays)

Reply:
Thanks. I added docstrings for this function.
But I can't find "eventseg.py", should I add it to /eventseg/event.py ?
But event.py contains a class, wouldn't it be strange to add this function to the class?

Reply:
My apologizes, that was a typo on my part (I was commenting on my own separate pull request at the same time) - I meant to say that this is specific to the *IS(F)C* code, so it should be moved to isfc.py. The utils file should be for functions that are expected to be useful in multiple modules, but I don't think this one is.

Reply:
I see. But I removed this function, so there's no problem now.